python/kwiver/sprokit/adapters/adapter_data_set.cxx

```diff
@@ -0,0 +1,296 @@
+  ADS_ADD_OBJECT(kwiver::vital::f2f_homography, kwiver::vital::f2f_homography)
+
+  #undef ADS_ADD_OBJECT
+
+  throw py::type_error("Unable to add object to adapter data set");
+}
+
+// Take data of an unknown type from a port and return. Can't return as an "any" object,
+// so need to cast
+py::object get_port_data_correct_type(adapter_data_set &self, ::sprokit::process::port_t const& port)
+{
+  kwiver::vital::any const any = self.get_port_data<kwiver::vital::any>(port);
+
+  #define ADS_GET_OBJECT(TYPE) \
+  if (any.is_type<TYPE>()) \
+  { \
+    return py::cast(kwiver::vital::any_cast<TYPE>(any)); \
+  }
+
+  ADS_GET_OBJECT(int)
+  ADS_GET_OBJECT(float)
+  ADS_GET_OBJECT(std::string)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::image_container>)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::descriptor_set>)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::detected_object_set>)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::track_set>)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::feature_track_set>)
+  ADS_GET_OBJECT(std::shared_ptr<kwiver::vital::object_track_set>)
+  ADS_GET_OBJECT(std::shared_ptr<std::vector<double>>)
+  ADS_GET_OBJECT(std::shared_ptr<std::vector<std::string>>)
+  ADS_GET_OBJECT(std::shared_ptr<std::vector<unsigned char>>)
+  ADS_GET_OBJECT(kwiver::vital::bounding_box_d)
+  ADS_GET_OBJECT(kwiver::vital::timestamp)
+  ADS_GET_OBJECT(kwiver::vital::geo_polygon)
+  ADS_GET_OBJECT(kwiver::vital::f2f_homography)
+
+  #undef ADS_GET_OBJECT
+
+  std::string msg("Unable to convert object found at adapter data set port: ");
+  msg += port;
+  msg += ". Data is of type: ";
+  msg += any.type_name();
+  throw py::type_error(msg);
+}
+
+}
+}
+}
+
+
+PYBIND11_MODULE(adapter_data_set, m)
+{
+  py::bind_vector<std::vector<unsigned char>, std::shared_ptr<std::vector<unsigned char>>>(m, "VectorUChar", py::module_local(true));
+  py::bind_vector<std::vector<double>, std::shared_ptr<std::vector<double>>>(m, "VectorDouble", py::module_local(true));
+  py::bind_vector<std::vector<std::string>, std::shared_ptr<std::vector<std::string>>>(m, "VectorString", py::module_local(true));
+
+  py::enum_<adapter_data_set::data_set_type>(m, "DataSetType"
+    , "Type of data set.")
+    .value("data", adapter_data_set::data)
+    .value("end_of_input", adapter_data_set::end_of_input)
+  ;
+
+  py::class_< adapter_data_set, std::shared_ptr<adapter_data_set > > ads(m, "AdapterDataSet");
+    ads.def_static("create", &adapter_data_set::create
+        , (py::arg("type") = adapter_data_set::data_set_type::data))
+
+    .def("__iter__", [](adapter_data_set &self)
+    {
+      return py::make_iterator(self.cbegin(), self.cend());
+    }, py::keep_alive<0,1>())
+
+    // Members
+    .def("type", &adapter_data_set::type)
+    .def("is_end_of_data", &adapter_data_set::is_end_of_data)
+
+    // General add_value which adds any type, and __setitem__
+    .def("add_value", &kwiver::sprokit::python::add_value_correct_type)
```

Comment:
This should have a docstring indicating it is equivalent to `__getitem__`, the same with the `get_port_data` function.

Reply:
How does a0322134e3f4a072f38fcba0b4da0bab9d227a23 look?